src/routes/v3/payouts.rs

```
@@ -0,0 +1,732 @@
+        }
+        _ => {}
+    }
+
+    Ok(HttpResponse::NoContent().finish())
+}
+
+#[get("")]
+pub async fn user_payouts(
+    req: HttpRequest,
+    pool: web::Data<PgPool>,
+    redis: web::Data<RedisPool>,
+    session_queue: web::Data<AuthQueue>,
+) -> Result<HttpResponse, ApiError> {
+    let user = get_user_from_headers(
+        &req,
+        &**pool,
+        &redis,
+        &session_queue,
+        Some(&[Scopes::PAYOUTS_READ]),
+    )
+    .await?
+    .1;
+
+    let payout_ids =
+        crate::database::models::payout_item::Payout::get_all_for_user(user.id.into(), &**pool)
+            .await?;
+    let payouts =
+        crate::database::models::payout_item::Payout::get_many(&payout_ids, &**pool).await?;
+
+    Ok(HttpResponse::Ok().json(
+        payouts
+            .into_iter()
+            .map(crate::models::payouts::Payout::from)
+            .collect::<Vec<_>>(),
+    ))
+}
+
+#[derive(Deserialize)]
+pub struct Withdrawal {
+    #[serde(with = "rust_decimal::serde::float")]
+    amount: Decimal,
+    method: PayoutMethodType,
+    method_id: String,
+}
+
+#[post("")]
+pub async fn create_payout(
+    req: HttpRequest,
+    pool: web::Data<PgPool>,
+    redis: web::Data<RedisPool>,
+    body: web::Json<Withdrawal>,
+    session_queue: web::Data<AuthQueue>,
+    payouts_queue: web::Data<PayoutsQueue>,
+) -> Result<HttpResponse, ApiError> {
+    let (scopes, user) =
+        get_user_record_from_bearer_token(&req, None, &**pool, &redis, &session_queue)
+            .await?
+            .ok_or_else(|| ApiError::Authentication(AuthenticationError::InvalidCredentials))?;
+
+    if !scopes.contains(Scopes::PAYOUTS_WRITE) {
+        return Err(ApiError::Authentication(
+            AuthenticationError::InvalidCredentials,
+        ));
+    }
+
+    let mtx = payouts_queue.lock_user_payouts(user.id.into());
+    let _guard = mtx.lock().await;
+
+    if user.balance < body.amount || body.amount < Decimal::ZERO {
```

I dont think `body.amount < Decimal::ZERO` should return this error - maybe make it use the 'iwth draw more to cover the fee' one below?

---

If the body amount is negative it's invalid, this catches this